main/src/ca/uwaterloo/flix/language/phase/Stratifier.scala

```diff
@@ -422,255 +424,264 @@ object Stratifier extends Phase[Root, Root] {
-        case (acc, SelectChannelRule(_, exp1, exp2)) => acc + dependencyGraphOfExp(exp1) + dependencyGraphOfExp(exp2)
+        case (acc, SelectChannelRule(_, exp1, exp2)) => acc + constraintGraphOfExp(exp1) + constraintGraphOfExp(exp2)
       }
 
     case Expression.Spawn(exp, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.Lazy(exp, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.Force(exp, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.FixpointConstraintSet(cs, _, _, _) =>
-      cs.foldLeft(DependencyGraph.empty) {
-        case (dg, c) => dg + dependencyGraphOfConstraint(c)
+      cs.foldLeft(ConstraintGraph.empty) {
+        case (dg, c) => dg + constraintGraphOfConstraint(c)
       }
 
     case Expression.FixpointMerge(exp1, exp2, _, _, _, _) =>
-      dependencyGraphOfExp(exp1) + dependencyGraphOfExp(exp2)
+      constraintGraphOfExp(exp1) + constraintGraphOfExp(exp2)
 
     case Expression.FixpointSolve(exp, _, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.FixpointFilter(_, exp, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.FixpointProjectIn(exp, _, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.FixpointProjectOut(_, exp, _, _, _) =>
-      dependencyGraphOfExp(exp)
+      constraintGraphOfExp(exp)
 
     case Expression.Reify(_, _, _, _) =>
-      DependencyGraph.empty
+      ConstraintGraph.empty
 
     case Expression.ReifyType(_, _, _, _, _) =>
-      DependencyGraph.empty
+      ConstraintGraph.empty
 
     case Expression.ReifyEff(_, exp1, exp2, exp3, _, _, _) =>
-      dependencyGraphOfExp(exp1) + dependencyGraphOfExp(exp2) + dependencyGraphOfExp(exp3)
+      constraintGraphOfExp(exp1) + constraintGraphOfExp(exp2) + constraintGraphOfExp(exp3)
   }
 
   /**
-    * Returns the dependency graph of the given constraint `c0`.
+    * Returns the constraint graph of the given constraint `c0`.
     */
-  private def dependencyGraphOfConstraint(c0: Constraint): DependencyGraph = c0 match {
+  private def constraintGraphOfConstraint(c0: Constraint): ConstraintGraph = c0 match {
     case Constraint(_, head, body, _) =>
       getPredicate(head) match {
-        case None => DependencyGraph.empty
+        case None => ConstraintGraph.empty
         case Some(headSym) =>
-          val dependencies = body flatMap (b => visitDependencyEdge(headSym, b))
-          DependencyGraph(dependencies.toSet)
+          val (pos, neg) = body.foldLeft((Set.empty[(Name.Pred, SourceLocation)], Set.empty[(Name.Pred, SourceLocation)])) {
+            case ((pos, neg), b) => b match {
+              case Body.Atom(pred, _, polarity, _, _, loc) => polarity match {
+                case Polarity.Positive => (pos + ((pred, loc)), neg)
```

Comment:
I guess the same predicate could occur twice, e.g. A :- B, B. But we probably dont care that it then occurs twice in the constraint graph. Its probably fine.

Reply:
Oh yeah I forgot about the location in the tuple. There isn't really any reason to have them be sets when the location is considered for equality?